ambry-replication/src/main/java/com/github/ambry/replication/ReplicaThread.java

```diff
@@ -1892,12 +2069,332 @@ void shutdown() throws InterruptedException {
+        DataNodeId remoteNode = remoteReplicas.get(0).getReplicaId().getDataNodeId();
+
+        // get stand by group id for the data node
+        int standByGroupId = standByGroupIds.get(remoteNode);
+
+        List<RemoteReplicaInfo> activeReplicas = new ArrayList<>();
+        List<RemoteReplicaInfo> standbyReplicasWithNoProgress = new ArrayList<>();
+
+        // generate active and standby replicas with no progress
+        fillActiveAndStandByNoProgressReplicas(activeReplicas, standbyReplicasWithNoProgress, remoteReplicas,
+            standByGroupId, groupId);
+        logger.trace("Thread name: {} Active replicas are {}, standby Replicas with no progress are {}", threadName,
+            activeReplicas, standbyReplicasWithNoProgress);
+
+        // if the current group is not in iteration, and there are active replicas, we will start a new group
+        // and increase iteration count if we are able to create
+        if (!inflightGroups.containsKey(groupId)) {
+          if (!activeReplicas.isEmpty()) {
+            RemoteReplicaGroup groupForActiveReplicas =
+                new RemoteReplicaGroup(activeReplicas, remoteNode, false, groupId);
+            inflightGroups.put(groupId, groupForActiveReplicas);
+            groupIdIterationCountMap.put(groupId, groupIdIterationCountMap.getOrDefault(groupId, 0) + 1);
+            logger.trace("Thread name: {} Created new group for groupId {}, iteration count is {}", threadName, groupId,
+                groupIdIterationCountMap.getOrDefault(groupId, 0));
+          }
+        }
+
+        // get timed out stand by replicas with no progress
+        List<RemoteReplicaInfo> standByReplicasTimedOutOnNoProgress =
+            getRemoteStandbyReplicasTimedOutOnNoProgress(standbyReplicasWithNoProgress);
+
+        if (!standByReplicasTimedOutOnNoProgress.isEmpty()) {
+          // add timed out stand by replicas with no progress to queue
+          standByReplicaQueue.computeIfAbsent(remoteNode, (r -> new HashSet<>()))
+              .addAll(standByReplicasTimedOutOnNoProgress);
+          logger.trace("Thread name: {} Added standby no progress replicas {} to queue", threadName,
+              standByReplicasTimedOutOnNoProgress);
+        }
+      }
+    }
+
+    /**
+     * iterate over all data nodes and stand by queue of all data nodes {@link #standByReplicaQueue},
+     * and tries to create standBy Groups and add to {@link #inflightGroups}
+     */
+    private void createStandByGroups() {
+      logger.trace("Thread name: {} Trying to create new standby groups", threadName);
+      for (Map.Entry<DataNodeId, Set<RemoteReplicaInfo>> entry : standByReplicaQueue.entrySet()) {
+        DataNodeId remoteNode = entry.getKey();
+        List<RemoteReplicaInfo> remoteReplicaInfos = new ArrayList<>(entry.getValue());
+        int groupId = standByGroupIds.get(remoteNode);
+        logger.trace("Thread name: {} Trying for remoteNode {}, standBy groupId {}}", threadName, remoteNode, groupId);
+
```

Comment:
if the standby replicas here are always timed-out, will their replica-group be marked as DONE somehwere ? because only if the group is marked Done, the code can move fwd here right ?

Reply:
each data node has only one timed out standby group. When this timed out stand by group finishes, only then, we can create a new group for timedout standby replicas in queue.

Current code also has only one standby group, I am assuming that is being done to reduce cross colo bandwidth usage.